docs/spec.md

```diff
@@ -1372,27 +1372,32 @@ dimensions `k` in `operand`.
 
 #### Inputs
 
-| Label | Name                   | Type                                         | Constraints   |
-|-------|------------------------|----------------------------------------------|---------------|
-| (I1)  | `operand`              | tensor                                       | (C1-C3), (C5) |
-| (I2)  | `broadcast_dimensions` | 1-dimensional tensor constant of type `si64` | (C2-C5)       |
+| Label | Name                   | Type                                         | Constraints      |
+|-------|------------------------|----------------------------------------------|------------------|
+| (I1)  | `operand`              | tensor or quantized tensor                   | (C1-C2), (C4-C8) |
+| (I2)  | `broadcast_dimensions` | 1-dimensional tensor constant of type `si64` | (C1-C4)          |
 
 #### Outputs
 
-| Name     | Type   | Constraints      |
-|----------|--------|------------------|
-| `result` | tensor | (C1), (C3), (C5) |
+| Name     | Type                       | Constraints   |
+|----------|----------------------------|---------------|
+| `result` | tensor or quantized tensor | (C2), (C4-C7) |
 
 #### Constraints
 
-* (C1) `operand` and `result` have the same element type.
-* (C2) size(`broadcast_dimensions`) $=$ rank(`operand`).
-* (C3) $0 \le$ `broadcast_dimensions[i]` $\lt$ rank(`result`) for all
+* (C1) size(`broadcast_dimensions`) $=$ rank(`operand`).
+* (C2) $0 \le$ `broadcast_dimensions[i]` $\lt$ rank(`result`) for all
        dimensions i in `operand`.
-* (C4) All dimensions in `broadcast_dimensions` are unique.
-* (C5) For all dimensions `j` in `operand`:
+* (C3) All dimensions in `broadcast_dimensions` are unique.
+* (C4) For all dimensions `j` in `operand`:
   * `dim(operand, j) = 1` or
   * `dim(operand, j) = dim(result, i)` where `i = broadcast_dimensions[j]`.
+* If the operation uses non-quantized tensors:
+  * (C5) `operand` and `result` have the same element type.
+* If the operation uses quantized tensors:
+  * (C6) `is_quantized_tensor(operand) and is_quantized_tensor(result)`.
+  * (C7) `quantized_element_type(operand) = quantized_element_type(result)`.
+  * (C8) `is_empty(quantization_dimension(operand)`.
```

Comment:
Same feedback as in #1566: let's minimize the changes to the spec by: 1) using `baseline_type` and/or `baseline_element_type` where appropriate instead of branching on whether the operation uses or doesn't use quantized tensors, 2) pushing is_empty into types, as in "tensor or per-tensor quantized tensor".

Reply:
2) done for all the ops. 

Re:1

I agree that it would be great if we could avoid branching on qunatized type. As we discussed in https://github.com/openxla/stablehlo/pull/1566#discussion_r1220364289, we need to somehow merge  `element_type` and `quantized_element_type` as they are related to differnet non-terminals  `TensorType` and `QuantizedTensorType` resp. 

A few options that I can think of: 
1. `element_type(operand) = element_type(result)` if `is_quantized(operand)`, `quantized_element_type(operand) = quantized_element_type(result)` otherwise. (Not a big fan for this as this seems too verbose).

2. Similar to how `type(x)` refers to the type of both tensor or quantized tensor types, can we use `element_type` for the element types of both. 

Specifically,

We can have a dedicated non-terminal to represent the element-type of Tensor type. Like
```
TensorType ::= 'tensor' '<' Shape TensorElementType '>'
``` 

With that we can reserve the function name `element_type` to be defined as
```
def element_type(x: Value | Placeholder | Type):
 if type(x) == TensorType:
    return tensor_element_type(x)
  if type(x) == QuantizedTensorType:
    return quantized_element_type(x)
  if type(x) is not Type:
    return element_type(type(x))
```

An additional benefit with the above approach would be that we can skip constraints like `(C6) is_quantized_tensor(operand) and is_quantized_tensor(result)`.

Please let me know your opinion.

Reply:
Originally, I thought that we shouldn't introduce `element_type` at all, so that each potential callsite is explicit about the distinction between element types and quantized element types. However, upon further thinking, I am now convinced that this will be extra verbose without real benefits.

Overall, I agree with the proposal, except that if we rename ElementType to TensorElementType, then we should also rename QuantizedElementType to QuantizedTensorElementType.

How about doing this refactoring in a separate PR, which would also replace usages of `quantized_element_type` with `element_type` for ConvolutionOp and PadOp? I'll quickly review, we'll merge and then you'll be able to rebase #1566 (elementwise), #1535 (data movement 1) and this PR (data movement 2)?

Reply:
makes sense. [Here](https://github.com/openxla/stablehlo/pull/1629) is the proposed PR.